arch/xtensa/src/esp32s3/esp32s3_ble_adapter.c

```diff
@@ -0,0 +1,2710 @@
+
+  do
+    {
+      /* set default values for global states or resources */
+
+      g_lp_cntl.val = 0;
+      g_lp_stat.val = 0;
+      g_wakeup_req_sem = NULL;
+      g_btdm_slp_tmr = NULL;
+
+      /* configure and initialize resources */
+
+      g_lp_cntl.enable = (cfg->sleep_mode == ESP_BT_SLEEP_MODE_1) ? 1 : 0;
+      g_lp_cntl.no_light_sleep = 0;
+
+      if (g_lp_cntl.enable)
+        {
+#if CONFIG_MAC_BB_PD
+          if (!btdm_deep_sleep_mem_init())
+            {
+              goto error;
+            }
+
+          g_lp_cntl.mac_bb_pd = 1;
+#endif
+#ifdef CONFIG_PM
+          g_lp_cntl.wakeup_timer_required = true;
+#endif
+          /* async wakeup semaphore for VHCI */
+
+          g_wakeup_req_sem = semphr_create_wrapper(1, 0);
+          if (g_wakeup_req_sem == NULL)
+            {
+              goto error;
+            }
+
+          btdm_vnd_offload_task_register(BTDM_VND_OL_SIG_WAKEUP_TMR,
+                                        btdm_sleep_exit_phase0);
+        }
+
+      if (g_lp_cntl.wakeup_timer_required)
+        {
+          esp_timer_create_args_t create_args =
+            {
+              .callback = btdm_slp_tmr_callback,
+              .arg = NULL,
+              .name = "btSlp",
+            };
+
+          if (esp_timer_create(&create_args, &g_btdm_slp_tmr) != OK)
+            {
+              wlerr("Failed to create timer");
+              goto error;
+            }
+        }
+
+      g_btdm_lpcycle_us_frac = RTC_CLK_CAL_FRACT;
+      g_btdm_lpcycle_us = 2 << (g_btdm_lpcycle_us_frac);
+
+      /* set default bluetooth sleep clock source */
+
+      g_lp_cntl.lpclk_sel = BTDM_LPCLK_SEL_XTAL; /* set default value */
+
+#if CONFIG_BT_CTRL_LPCLK_SEL_EXT_32K_XTAL
+      if (esp32s3_rtc_clk_slow_freq_get() == RTC_SLOW_FREQ_32K_XTAL)
+        {
+          g_lp_cntl.lpclk_sel = BTDM_LPCLK_SEL_XTAL32K;
+        }
+      else
+        {
+          wlwarn("32.768kHz XTAL not detected");
+#if !CONFIG_BT_CTRL_MAIN_XTAL_PU_DURING_LIGHT_SLEEP
+          g_lp_cntl.lpclk_sel = BTDM_LPCLK_SEL_XTAL;
+#endif
+        }
+#elif (CONFIG_BT_CTRL_LPCLK_SEL_MAIN_XTAL)
+      wlwarn("Bluetooth will use main XTAL as Bluetooth sleep clock.");
+#if !CONFIG_BT_CTRL_MAIN_XTAL_PU_DURING_LIGHT_SLEEP
+      g_lp_cntl.no_light_sleep = 1;
+#endif
+#elif (CONFIG_BT_CTRL_LPCLK_SEL_RTC_SLOW)
+      /* check whether or not internal 150 kHz RC oscillator is working */
+
+      if (rtc_clk_slow_src_get() == RTC_SLOW_FREQ_RTC)
+        {
+          g_lp_cntl.lpclk_sel = BTDM_LPCLK_SEL_RTC_SLOW; /* Internal 150 kHz RC oscillator */
+          wlwarn("Internal 150kHz RC osciallator. The accuracy of this "
+                 "clock is a lot larger than 500ppm which is required in "
+                 "Bluetooth communication, so don't select this option in "
+                 "scenarios such as BLE connection state.");
+        }
+      else
+        {
+          wlerr("Internal 150kHz RC oscillator not detected.");
+          assert(0);
```

Comment:
maybe `PANIC();` in kernel code instead of `assert(0);`? It has major difference on not relying on `NDEBUG` of course. This is just a topic for discussion and definitely a part of a separate PR.

Reply:
I totally agree and I think it fits on this same PR ;)